### PR TITLE
[APR] Implement API route for round and apply it on frontend

### PR DESCRIPTION
**Pull request by Rawallon**
First render still takes ~1m to fetch data, but subsequent requests are cached

https://github.com/bleu-studio/balancer-tools/assets/68453900/77eec751-2e39-49b6-99ca-c29749ac3349

### CODE REVIEW

File: apps/balancer-tools/src/app/apr/(components)/KPI.tsx [hunk 1]
```diff
@@ -1,14 +1,34 @@
...
+        <KPICardWrapper>
+          <Spinner size="sm" />
+        </KPICardWrapper>
+      }
+    >
+      <KPICardWrapper>
+        <div className="font-semibold">{title}</div>
+        <div className="pt-2">{content}</div>
+      </KPICardWrapper>
+    </Suspense>
```
Comment (ribeirojose) [inline apps/balancer-tools/src/app/apr/(components)/KPI.tsx:32]:
```suggestion
    <KPICardWrapper>
    <Suspense
      fallback={
          <Spinner size="sm" />
      }
    >
        <div className="font-semibold">{title}</div>
        <div className="pt-2">{content}</div>
      </Suspense>
    </KPICardWrapper>
```

File: apps/balancer-tools/src/app/apr/(utils)/calculatePoolStats.ts [hunk 2]
```diff
@@ -9,6 +11,14 @@ import { Round } from "./rounds";
 
 const WEEKS_IN_YEAR = 52;
 
+// Whenever we exceed the subgraph rate limit, it would just stop returning data
+// p-throttle is used to limit the number of requests to the subgraph at once
+// I got to these numbers by testing and adjusting manually
```
Comment (ribeirojose) [inline apps/balancer-tools/src/app/apr/(utils)/calculatePoolStats.ts:16]: 👍

File: apps/balancer-tools/src/app/apr/(utils)/calculatePoolStats.ts [hunk 3]
```diff
@@ -19,12 +29,31 @@ export async function calculatePoolStats({
...
+      apr: -1,
+      balPriceUSD: -1,
+      tvl: -1,
+      votingShare: -1,
+      symbol: -1,
+      network: -1,
+    };
+  }
 
+  const network = pool.gauge?.network ?? 1;
```
Comment (ribeirojose) [inline apps/balancer-tools/src/app/apr/(utils)/calculatePoolStats.ts:56]: can you rebase this PR on mine? #239

File: apps/balancer-tools/src/app/apr/api/route.ts [hunk 4]
```diff
@@ -5,26 +5,92 @@ import votingGauges from "#/data/voting-gauges.json";
...
     const gauge = votingGauges.filter((gauge) => gauge.pool.id === poolId)[0];
     // Multiplying by 1000 because unix timestamp is in seconds
     const gaugeAddedDate = new Date(gauge["addedTimestamp"] * 1000);
-    const roundGaugeAdded = Round.getRoundByDate(gaugeAddedDate);
+    let roundGaugeAdded = Round.getRoundByDate(gaugeAddedDate);
+    if (Round.getRoundByNumber(1).endDate > gaugeAddedDate) {
+      // There are pools that were added before the first round, so we need to handle that
+      // Ex: 0x5c6ee304399dbdb9c8ef030ab642b10820db8f56000200000000000000000014
+      roundGaugeAdded = Round.getRoundByNumber(1);
+    }
```
Comment (ribeirojose) [inline apps/balancer-tools/src/app/apr/api/route.ts:91]: You could just tweak `getRoundByDate` to return 1 in that case as well

File: apps/balancer-tools/src/app/apr/pool/(components)/HistoricalAPRChart.tsx [hunk 5]
```diff
@@ -14,13 +15,19 @@ export default async function HistoricalAPRChart({
...
-  const results = await generatePromisesForHistoricalPoolData(poolId);
-  results.map((result) => {
-    APRPerRoundCords.x.push(`Round ${result.roundId}`);
-    APRPerRoundCords.y.push(result.apr);
-  });
+  const results: PoolStatsResults = await fetcher(
+    `${process.env.NEXT_PUBLIC_SITE_URL}/apr/api/?poolid=${poolId}`,
+  );
+
+  const APRPerRoundCords = Object.entries(results.perRound).reduce(
```
Comment (ribeirojose) [inline apps/balancer-tools/src/app/apr/pool/(components)/HistoricalAPRChart.tsx:23]:
```suggestion
  const aprPerRoundCords = Object.entries(results.perRound).reduce(
```

File: apps/balancer-tools/src/app/apr/round/(components)/TopPoolsChart.tsx [hunk 6]
```diff
@@ -41,26 +21,21 @@ export default async function TopPoolsChart({ roundId }: { roundId: string }) {
...
-    y: string[];
-    colors: string[];
-  } = {
-    x: results.map((result) => result.apr.toFixed(2)),
-    y: results.map((result) => result.symbol),
-    colors: results.map((_, index) => colors[index % colors.length]),
-  };
+  const topAprApi: PoolStatsData[] = Object.values(
+    await fetcher(
+      `${process.env.NEXT_PUBLIC_SITE_URL}/apr/api/?roundid=${roundId}&sort=apr&limit=10`,
```
Comment (ribeirojose) [inline apps/balancer-tools/src/app/apr/round/(components)/TopPoolsChart.tsx:26]: 👌 but you probably need sort direction as well

File: apps/balancer-tools/src/app/apr/round/(components)/PoolListTable.tsx [hunk 7]
```diff
@@ -119,26 +124,37 @@ export function PoolListTable({ roundId }: { roundId: string }) {
...
+                </Button>
+              ) : (
+                <Button
+                  className="w-full flex content-center justify-center gap-x-3 rounded-t-none rounded-b disabled:cursor-not-allowed"
+                  shade="medium"
+                  onClick={loadMorePools}
+                >
+                  Load More <ChevronDownIcon />
+                </Button>
+              )}
```
Comment (ribeirojose) [inline apps/balancer-tools/src/app/apr/round/(components)/PoolListTable.tsx:157]: 👌 great UI enhancement

File: apps/balancer-tools/src/app/apr/(components)/KPI.tsx (hunk 1 above)
Comment (Rawallon) [reply apps/balancer-tools/src/app/apr/(components)/KPI.tsx:32]: 🤯

Não sei como consegui pensar no pior maneira possível de se fazer isso hahaha

**Comment by ribeirojose** [comment]
Definitely in on the right track -- the API implementation is almost there, my point is that we should recursively cache the result for all pools/rounds combination. I added a draft PR with adding that idea (not functional) and some refactors using your PR as base https://github.com/bleu-studio/balancer-tools/pull/240. let me know if you have any problems understanding any part of that!

**Pull request merged**